File: orchestrator/run_agent.py
"""
orchestrator/run_agent.py — Core function to run a single LLM agent.

Loads the system prompt from agents/{role}.prompt.txt, calls the OpenAI-compatible
API, parses the JSON response, and retries on failure.
"""

import json
import os
import time
from pathlib import Path

import requests

from utils.parse_json import parse_json
from utils.agent_logger import log_agent_output

AGENTS_DIR = Path(__file__).resolve().parent.parent / "agents"
MAX_RETRIES = 2


def run_agent(role: str, user_input: str) -> dict:
    """
    Run a single agent: load prompt, call LLM, parse JSON response.

    Args:
        role: Agent name (architect|product|engineer|qa)
        user_input: The user message / context to send

    Returns:
        Parsed JSON dict from the agent
    """
    api_key = os.getenv("OPENAI_API_KEY")
    if not api_key:
        raise RuntimeError("OPENAI_API_KEY environment variable is not set")

    base_url = os.getenv("OPENAI_BASE_URL", "https://api.openai.com/v1").rstrip("/")
    model = os.getenv("OPENAI_MODEL", "gpt-4o")

    prompt_path = AGENTS_DIR / f"{role}.prompt.txt"
    if not prompt_path.exists():
        raise FileNotFoundError(f"Prompt file not found: {prompt_path}")

    system_prompt = prompt_path.read_text(encoding="utf-8").strip()

    last_error = None

    for attempt in range(MAX_RETRIES + 1):
        try:
            response = requests.post(
                f"{base_url}/chat/completions",
                headers={
                    "Content-Type": "application/json",
                    "Authorization": f"Bearer {api_key}",
                },
                json={
                    "model": model,
                    "temperature": 0.2,
                    "messages": [
                        {"role": "system", "content": system_prompt},
                        {"role": "user", "content": user_input},
                    ],
                    "response_format": {"type": "json_object"},
                },
                timeout=120,
            )

            if response.status_code != 200:
                raise RuntimeError(
                    f"API returned {response.status_code}: {response.text[:500]}"
                )

            data = response.json()
            content = data.get("choices", [{}])[0].get("message", {}).get("content")

            if not content:
                raise RuntimeError("Empty response from LLM")

            parsed = parse_json(content)

            log_agent_output(role, {
                "attempt": attempt + 1,
                "model": model,
                "input": user_input[:200],
                "output": parsed,
                "timestamp": time.strftime("%Y-%m-%dT%H:%M:%S"),
            })

            return parsed

        except Exception as err:
            last_error = err
            is_last = attempt == MAX_RETRIES
            print(f"  [{role}] Attempt {attempt + 1}/{MAX_RETRIES + 1} failed: {err}")
            if not is_last:
                time.sleep(1 * (attempt + 1))

    raise RuntimeError(
        f'Agent "{role}" failed after {MAX_RETRIES + 1} attempts: {last_error}'
    )

File: orchestrator/runAgent.js
const fs = require("fs");
const path = require("path");
const { parseJSON } = require("../utils/parseJSON");
const { logAgentOutput } = require("../utils/logger");

const AGENTS_DIR = path.join(__dirname, "..", "agents");
const MAX_RETRIES = 2;

/**
 * Run a single agent: load its prompt, call the LLM, parse JSON response.
 * @param {string} role - Agent name (architect|product|engineer|qa)
 * @param {string} userInput - The user message / context to send
 * @returns {Promise<object>} Parsed JSON from the agent
 */
async function runAgent(role, userInput) {
  const apiKey = process.env.OPENAI_API_KEY;
  if (!apiKey) {
    throw new Error("OPENAI_API_KEY environment variable is not set");
  }

  const baseUrl = process.env.OPENAI_BASE_URL || "https://api.openai.com/v1";
  const model = process.env.OPENAI_MODEL || "gpt-4o";

  const promptPath = path.join(AGENTS_DIR, `${role}.prompt.txt`);
  if (!fs.existsSync(promptPath)) {
    throw new Error(`Prompt file not found: ${promptPath}`);
  }

  const systemPrompt = fs.readFileSync(promptPath, "utf-8").trim();

  let lastError = null;

  for (let attempt = 0; attempt <= MAX_RETRIES; attempt++) {
    try {
      const response = await fetch(`${baseUrl}/chat/completions`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${apiKey}`,
        },
        body: JSON.stringify({
          model,
          temperature: 0.2,
          messages: [
            { role: "system", content: systemPrompt },
            { role: "user", content: userInput },
          ],
          response_format: { type: "json_object" },
        }),
      });

      if (!response.ok) {
        const errorBody = await response.text();
        throw new Error(
          `API returned ${response.status}: ${errorBody.slice(0, 500)}`
        );
      }

      const data = await response.json();
      const content = data.choices?.[0]?.message?.content;

      if (!content) {
        throw new Error("Empty response from LLM");
      }

      const parsed = parseJSON(content);

      await logAgentOutput(role, {
        attempt: attempt + 1,
        model,
        input: userInput.slice(0, 200),
        output: parsed,
        timestamp: new Date().toISOString(),
      });

      return parsed;
    } catch (err) {
      lastError = err;
      const isLastAttempt = attempt === MAX_RETRIES;
      console.error(
        `[${role}] Attempt ${attempt + 1}/${MAX_RETRIES + 1} failed: ${err.message}`
      );
      if (!isLastAttempt) {
        const delay = 1000 * (attempt + 1);
        await new Promise((resolve) => setTimeout(resolve, delay));
      }
    }
  }

  throw new Error(
    `Agent "${role}" failed after ${MAX_RETRIES + 1} attempts: ${lastError?.message}`
  );
}

module.exports = { runAgent };
